test/reader_test.go

```diff
@@ -0,0 +1,556 @@
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/writeconcern"
+	"go.uber.org/zap"
+)
+
+var dummyKv []model.KeyValue = []model.KeyValue{
+	{
+		Key:    "http.status_code",
+		VType:  model.Int64Type,
+		VInt64: 200,
+	},
+}
+
+var statusCode404 []model.KeyValue = []model.KeyValue{
+	{
+		Key:    "http.status_code",
+		VType:  model.Int64Type,
+		VInt64: 404,
+	},
+}
+
+// Returns the non-loopback local IP address of the host
+func GetLocalIP() string {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return ""
+	}
+	for _, address := range addrs {
+		// if it is not a loopback ip address, use it
+		if ipnet, ok := address.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+			if ipnet.IP.To4() != nil {
+				return ipnet.IP.String()
+			}
+		}
+	}
+	return ""
+}
+
+// Helper function to verify depedency pattern is valid.
+func isValidDepedencyPattern(pattern string) bool {
+	switch pattern {
+	case
+		"none",
+		"single",
+		"circular":
+		return true
+	}
+	return false
+}
+
+// Helper function to generate traces for testing purposes.
+func GenerateTraces(ctx context.Context, writer *jaeger_mongodb.SpanWriter, numTraces int, dependencyPattern string, followsFrom bool) {
+	if !isValidDepedencyPattern(dependencyPattern) {
+		zap.S().Fatal(fmt.Printf("Error: Dependency pattern '%s' does not exist!\n", dependencyPattern))
+	}
+	randomOperation := map[int]string{0: "grpc", 1: "http", 2: "spark", 3: "redis"}
+	for i := 0; i < numTraces; i++ {
+		tags := dummyKv
+		if i%10 == 0 {
+			tags = statusCode404
+		}
+		s := model.Span{
+			TraceID:       model.NewTraceID(uint64(i), uint64(i)),
+			SpanID:        model.NewSpanID(uint64(i)),
+			OperationName: randomOperation[i%4],
+			References:    []model.SpanRef{},
+			StartTime:     time.Now().Add(time.Duration(i)),
+			Duration:      time.Duration(i + 10),
+			Tags:          tags,
+			Process: &model.Process{
+				ServiceName: fmt.Sprintf("Service %d", i),
+				Tags:        dummyKv,
+			},
+			Logs: []model.Log{
+				{
+					Timestamp: time.Now(),
+					Fields:    dummyKv,
+				},
+			},
+		}
+		if dependencyPattern == "circular" && i < numTraces-1 {
+			newRef := model.NewChildOfRef(model.TraceID{Low: uint64(i + 1), High: uint64(i + 1)}, model.SpanID(uint64(i+1)))
+			if followsFrom {
+				newRef = model.NewFollowsFromRef(model.TraceID{Low: uint64(i + 1), High: uint64(i + 1)}, model.SpanID(uint64(i+1)))
+			}
+			s.References = append(s.References, newRef)
+		}
+		if (dependencyPattern == "single" || dependencyPattern == "circular") && i > 0 {
+			newRef := model.NewChildOfRef(model.TraceID{Low: uint64(i - 1), High: uint64(i - 1)}, model.SpanID(uint64(i-1)))
+			if followsFrom {
+				newRef = model.NewFollowsFromRef(model.TraceID{Low: uint64(i - 1), High: uint64(i - 1)}, model.SpanID(uint64(i-1)))
+			}
+			s.References = append(s.References, newRef)
+		}
+		writer.WriteSpan(ctx, &s)
+	}
+}
+func TestReaderIntegration(t *testing.T) {
+	ip := GetLocalIP()
```

Comment:
instead of using this function we should gate the IT tests on an environment variable that will both provide the mongo url and then run the IT tests.

```
	mongoURL := os.Getenv("MONGO_URL")
	if mongoURL == "" {
		t.Skip("set mongoURL to run the IT tests")
	}

```

and then if we set `export MONGO_URL=mongodb://localhost:27017` it will connect to the mongo running locally.


https://peter.bourgon.org/blog/2021/04/02/dont-use-build-tags-for-integration-tests.html

Reply:
resolved, thanks